torch/_inductor/cudagraph_trees.py

```diff
@@ -634,47 +651,47 @@ def __init__(
                 self.outputs_metadata.append(None)
 
-        # initialized on first run
-        self.checkpointed_caching_state: Optional[AllocatorState] = None
+        self.graph.replay()
+
+    def _copy_input(self, idx, dst, src):
+        expanded_dims = self.expanded_dims[idx]
+        dst = index_expanded_dims(dst, expanded_dims)
+        src = index_expanded_dims(src, expanded_dims)
+        # TODO - one jit kernel across multiple inputs
+        dst.copy_(src)
 
     def run(self, new_inputs):
         if config.triton.slow_cudagraph_asserts:
             self.debug_check_invariants_before_invocation()
 
-        assert len(self.static_input_data_ptrs) == len(new_inputs)
-
         storage_cache = {}
-        # NB: this ranges over non-static inputs too
-        for idx, data_ptr in enumerate(self.static_input_data_ptrs):
-            if idx in self.cudagraph_managed_idxs:
-                continue
-            if data_ptr is not None:
-                # static input, e.g., parameter
-                assert data_ptr == new_inputs[idx].data_ptr()
-            else:
-                # non-static input, need to copy it into CUDA graph
-                dst = self._reconstruct_from_tensor_metadata(
-                    self.non_static_inputs_metadata[idx], storage_cache
-                )
-                src = new_inputs[idx]
-                expanded_dims = self.expanded_dims[idx]
-
-                dst = index_expanded_dims(dst, expanded_dims)
-                src = index_expanded_dims(src, expanded_dims)
-                # TODO - one jit kernel across multiple inputs
-                dst.copy_(src)
 
-        new_inputs.clear()
-        self.graph.replay()
-
-        # outputs is not None on first execution
-        if self.recording_outputs is not None:
+        if self.recording_outputs is None:
+            assert len(self.static_input_data_ptrs) == len(new_inputs)
+            # NB: this ranges over non-static inputs too
+            for idx, data_ptr in enumerate(self.static_input_data_ptrs):
+                if idx in self.cudagraph_managed_idxs:
+                    continue
+                if data_ptr is not None:
+                    # static input, e.g., parameter
+                    assert data_ptr == new_inputs[idx].data_ptr()
+                else:
+                    # non-static input, need to copy it into CUDA graph
+                    dst = self._reconstruct_from_tensor_metadata(
+                        self.non_static_inputs_metadata[idx], storage_cache
+                    )
+                    src = new_inputs[idx]
+                    self._copy_input(idx, dst, src)
+        else:
+            # inputs are copied over in _allocate_recording_inputs and subsequently cleared
+            assert len(new_inputs) == 0
             outputs = self.recording_outputs
             self.recording_outputs = None
-            self._add_first_outputs(outputs)
-
             return outputs
 
+        new_inputs.clear()
```

Comment:
You only need to do this in the if branch, right? Not the else.

Reply:
We're returning in the else